### ISSUE TITLE
ObjectDisposedException on Process.Start leaves defunct process on Linux

### ISSUE THREAD

**Issue by markmcdowell** [dotnet/corefx#27249]
This is a difficult one to reproduce! 

We're starting a process repeatedly in 5 seconds gaps. After leaving running for a week, we end up with 4-5 defunct processes that failed to start. 
On the ones that failed we see the process handle is marked as invalid and an exception (see below) is thrown. Even though we're disposing, the process now has a defunct child for it's lifetime.

C# details:

- Self contained deployment
- DotNetCore 2.0
- Centos 7.3

Code to reproduce:
```csharp
using (var process = new Process())
{
    process.StartInfo.FileName = "filename";
    process.StartInfo.Arguments = "arguments";
    process.StartInfo.RedirectStandardOutput = true;
    process.EnableRaisingEvents = true;
    process.Start();

    var output = process.StandardOutput.ReadToEnd();

    process.WaitForExit(MaxWaitTime);
}
```

Exception stack:
```csharp
System.ObjectDisposedException: Safe handle has been closed
at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
at System.Threading.RegisteredWaitHandleSafe.SetWaitObject(WaitHandle waitObject)
at System.Threading.ThreadPool.RegisterWaitForSingleObject(WaitHandle waitObject, WaitOrTimerCallback callBack, Object state, UInt32 millisecondsTimeOutInterval, Boolean executeOnlyOnce, StackCrawlMark& stackMark, Boolean compressStack)
at System.Threading.ThreadPool.RegisterWaitForSingleObject(WaitHandle waitObject, WaitOrTimerCallback callBack, Object state, Int32 millisecondsTimeOutInterval, Boolean executeOnlyOnce)
at System.Diagnostics.Process.EnsureWatchingForExit()
at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
at System.Diagnostics.Process.Start()
```

**Comment by danmosemsft** [comment]
cc @tmds since he has worked on a relevant area, albeit this is in 2.0 bits.

**Comment by tmds** [comment]
I try to reproduce this on dotnet 2.0.3 with this code, but no luck:

```C#
            for (int i = 0; i < 1_000_000; i++)
            {
                if (i % 100 == 0)
                {
                    System.Console.Write(".");
                }
                using (var process = new Process())
                {
                    process.StartInfo.FileName = "/usr/bin/sleep";
                    process.StartInfo.Arguments = "0";
                    process.StartInfo.RedirectStandardOutput = true;
                    process.EnableRaisingEvents = true;
                    process.Start();

                    var output = process.StandardOutput.ReadToEnd();

                    process.WaitForExit();
                }
            }
```

@markmcdowell What version of dotnet are you using exactly? 
Does the process you call terminate quick? Or does it run for several seconds at least?
How much is `MaxWaitTime` compared to the process expected duration? Multiple times larger?

**Comment by tmds** [comment]
After looking a bit in the code, it seems this could be due to pid recycling.
Updated reproducer:
```C#
            List<Process> processes = new List<Process>();
            for (int i = 0; i < 1_000_000; i++)
            {
                if (i % 100 == 0)
                {
                    System.Console.Write(".");
                }
                try
                {
                    var process = new Process();
                    processes.Add(process);
                    process.StartInfo.FileName = "/usr/bin/sleep";
                    process.StartInfo.Arguments = "0";
                    process.EnableRaisingEvents = true;
                    process.Start();

                    process.WaitForExit();
                }
                catch
                {
                    System.Console.WriteLine($"Exception when creating process {i}");
                    throw;
                }
            }
```
Which fails with:
```
Exception when creating process 25833

Unhandled Exception: System.ObjectDisposedException: Safe handle has been closed
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.Threading.RegisteredWaitHandleSafe.SetWaitObject(WaitHandle waitObject)
   at System.Threading.ThreadPool.RegisterWaitForSingleObject(WaitHandle waitObject, WaitOrTimerCallback callBack, Object state, UInt32 millisecondsTimeOutInterval, Boolean executeOnlyOnce, StackCrawlMark& stackMark, Boolean compressStack)
   at System.Threading.ThreadPool.RegisterWaitForSingleObject(WaitHandle waitObject, WaitOrTimerCallback callBack, Object state, Int32 millisecondsTimeOutInterval, Boolean executeOnlyOnce)
   at System.Diagnostics.Process.EnsureWatchingForExit()
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
```
So a new process is trying to a Disposed WaitHandle from an old process which had the same pid.

@markmcdowell this doesn't match exactly with your repro code. Are you keeping references to the Processes? Do you do something using the `Exited` event (since you set `EnableRaisingEvents`)?

**Comment by markmcdowell** [comment]
> What version of dotnet are you using exactly?

I believe the exact version is 2.0.0, in the deps.json file I see:
```
"runtimeTarget": {
    "name": ".NETCoreApp,Version=v2.0/centos.7-x64",
    "signature": "176cc939ab95dfc30e8a31d919625445d48fa978"
  },
...
"Microsoft.NETCore.App": "2.0.0",
```
> Does the process you call terminate quick? Or does it run for several seconds at least?

Yes it finishes within 1 second.

> How much is MaxWaitTime compared to the process expected duration? Multiple times larger?

Much larger at the moment, 10 mins

> Are you keeping references to the Processes? Do you do something using the Exited event (since you set EnableRaisingEvents)?

No the code posted is pretty exact, we dispose of the process straight away. We're not actually using the exited event (EnableRaisingEvents was added due to googling defunct processes in dotnet core).

We do have another thread creating processes as well, would both running at the same time cause it to use the same pid?

**Comment by tmds** [comment]
> No the code posted is pretty exact, we dispose of the process straight away. We're not actually using the exited event (EnableRaisingEvents was added due to googling defunct processes in dotnet core).

Ah ok. I think if you remove `EnableRaisingEvents` and also remove the `MaxWaitTime` argument from `WaitForExit`, you should no longer see the issue.

> We do have another thread creating processes as well, would both running at the same time cause it to use the same pid?

It is possible, but not likely.
For this exception to happen, someone still needs to have a reference to the old process.
In my reproducer I keep references to the old Process and don't Dispose them.

With your code, this means one thread (old process) needs to be between `WaitPid` and `Process.Dispose`.
While on the other thread, the OS hands that pid to the new process in Process.Start.
It would be strange for the OS to recycle the pid so fast.

The implementation uses reference counting to share state (`ProcessWaitState.Holder`) between Process instances for the same pid. If the reference count would due to some bug not go to zero, we'll also get this exception for a pid that is recycled after a long time.

**Comment by markmcdowell** [comment]
Ok thanks, I'll remove `EnableRaisingEvents` and `MaxWaitTime` and run that for a while.

**Comment by danmosemsft** [comment]
@tmds does this repro after your recent changes to use waitpid? If so do you believe a fix is possible on our side?

At least on Windows, how PID's are recycled is to my knowledge completely undefined and sometimes changes.

**Comment by tmds** [comment]
@danmosemsft the issue I produced is a know limitation of the 2.0 implementation that it can't handle recycled pids if someone still has references to the old process. I'm not sure if that is a match with what @markmcdowell reported because in his code he is not keeping any references to the processes.

I think the first issue @markmcdowell had (before adding `EnableRaisingEvents`) was child processes not being reaped when they occasionally ran longer than `MaxWaitTime`. This is what is being fixed in https://github.com/dotnet/corefx/pull/26291.
Then by adding `EnableRaisingEvents` he hit another issue. This could be (very unlikely) pid recyling between the two threads creating processes. More likely is a bug in the reference counting which causes a ProcessWaitState to stay around. I looked at the code and saw one place to improve and a bug in my PR (https://github.com/dotnet/corefx/pull/26291/commits/d5e6785595a4192316d5d000d34f682cc2ca9e83). I've added a test to the PR that checks reference counting: https://github.com/dotnet/corefx/pull/26291/commits/2221ec6568f5296dc6a98ec5cad8ca2a7cca5dce.
I think (based on the changes in the PR) we can support child pids being recycled even when still referenced. I'd rather look at that in a separate PR. Then this should no longer throw: https://github.com/dotnet/corefx/issues/27249#issuecomment-367288949.

At that point, we may assume we fixed this issue but I don't think we are a 100% sure since we couldn't reproduce it with the code used by @markmcdowell.

**Comment by danmosemsft** [comment]
Thank you for the detailed analysis @tmds. After your PR goes in it would be nice if @markmcdowell was able to repeat his stress test with latest bits.

**Comment by markmcdowell** [comment]
Hello! I've been running without `EnableRaisingEvents` and `MaxWaitTime` but unfortunately had no improvement. I still get the defunct processes. 

Full code I'm using:

```
string output;
using (var process = new Process())
{
    process.StartInfo.FileName = FileName;
    process.StartInfo.Arguments = Arguments;
    process.StartInfo.RedirectStandardOutput = true;
    process.Start();

    output = process.StandardOutput.ReadToEnd();

    process.WaitForExit();
}
```

**Comment by tmds** [comment]
A defunct process is a process that wasn't cleaned up by the parent.
The 'WaitForExit' is the way cleanup happens here. So it is unexpected you still get those processes with this code.

Perhaps if you look at the processes (e.g. in /proc, or using ps), maybe you see something interesting about them? E.g. is 'dotnet' the parent for these processes? Anything interesting about their command-line? Their exit code? ...

This PR (https://github.com/dotnet/corefx/pull/26291) changes when&how dotnet reaps child processes. Once it is merged, it would be good if you could try and run with those changes.

**Comment by tmds** [comment]
@markmcdowell https://github.com/dotnet/corefx/pull/26291 is merged. You can get a daily build from the cli repo: https://dotnetcli.blob.core.windows.net/dotnet/Sdk/master/dotnet-sdk-latest-linux-x64.tar.gz. I hope you won't see any more zombies when using this version.

**Comment by danmosemsft** [comment]
There is a few days delay for corefx to reach the SDK builds

**Comment by tmds** [comment]
@danmosemsft indeed, the PR was merged some time ago already and I checked: it is part of the daily cli build.

**Comment by tmds** [comment]
@markmcdowell This is also in the just-released 2.1-preview2 (https://github.com/dotnet/core/blob/master/release-notes/download-archives/2.1.0-preview2-download.md).

**Comment by markmcdowell** [comment]
Great thanks. I'll get it running today and let you know.

**Comment by tmds** [comment]
@markmcdowell seen any zombies lately? 😄

**Comment by markmcdowell** [comment]
Sorry for the slow response. I've yet to test because we can't get our projects to publish with dotnet core 2.1 (even the official release).

We use msbuild because it still has msbuild tasks in the solution and I get loads of errors like:

```
The project was restored using Microsoft.NETCore.App version 2.0.0, but with current settings, version 2.0.7 would be used instead.
```

I'll update when I hopefully get it working!

**Comment by DaramG** [comment]
On the latest dotnet(2.1.4), I can still reproduce this bug with the above code.
This bug seems to be not completely fixed . Can you check it?

**Comment by stephentoub** [comment]
2.1.4 would be the version of the sdk. What's the version of the runtime you're using? i.e. what version is specified in your .csproj? What's the output of `dotnet --info`?

**Comment by DaramG** [comment]
The `TarggetFramework` in .csproj was `netcoreapp2.0` and here is the result of `dotnet --info` on my machine.
```
.NET Command Line Tools (2.1.4)

Product Information:
 Version:            2.1.4
 Commit SHA-1 hash:  5e8add2190

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /usr/share/dotnet/sdk/2.1.4/

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.5
  Build    : 17373eb129b3b05aa18ece963f8795d65ef8ea54

```

**Comment by tmds** [comment]
@DaramG you are using .NET Core 2.0 (Microsoft .NET Core Shared Framework Host). Please change your installation to include .NET Core 2.1 (www.dot.net/download). If both 2.0 and 2.1 are installed, a `netcoreapp2.0` will prefer to run on 2.0. If you change to `netcoreapp2.1`, your app will require 2.1+. To build a `netcoreapp2.1` application, you need the 2.1.300 sdk.

**Comment by juandrn** [comment]
I had a similar problem but before a socket was left open and lose by time out, I don't have an example but I'll upload one
